python/ray/util/sgd/torch/distributed_torch_runner.py

```diff
@@ -103,12 +83,39 @@ def _setup_training(self):
             schedulers=self.schedulers,
             use_fp16=self.use_fp16)
 
+    def _initialize_dataloaders(self):
+        super(DistributedTorchRunner, self)._initialize_dataloaders()
+
+        def with_sampler(loader):
```

Comment:
Just realized that we have to recreate the loader on every runner anyway, so maybe document that? The first question I asked before is answered by this. Also why don't we collect the `data_loader_args` on trainer so that we don't have to rely on pickling a weird `torch.DataLoader` object?

The question about the performance of creating dataloaders still technically stands, since it is currently very unclear that the user's dataloader is actually never going to be used (and on distributed only!). This will cause issues if the user returns a `DataLoader` subclass and expect its methods to be used, for example. The inconsistency between distributed and local runner is also very important to fix here, since most people will test locally and then rely on things to "just work".

Reply:
We don't pickle the DataLoader; it's initialized/created in `_initialize_dataloaders`. 

It's pretty hard to patch the sampler on an existing dataloader since there's quite a bit of logic in the initializer, and the loader creation is quite cheap.

https://pytorch.org/docs/stable/_modules/torch/utils/data/dataloader.html#DataLoader

Right now, I've protected this with `isinstance(DataLoader)`, so this code won't run for subclasses.

Reply:
Seems like all the logic is in checking that the sampler is valid. But there is indeed no reason to try and patch it since all the constructor does is argument validity checks + assigning class variables.

Thanks for clarifying, I think this is good now.

Reply:
Awesome!